nemo/collections/llm/gpt/data/core.py

```diff
@@ -77,25 +109,1334 @@ def create_sft_dataset(
+
+            # load meta info
+            with open(idx_fn + ".info", "rb") as fp:
+                idx_info_dict = pickle.load(fp)
+            # test for mismatch in expected newline_int
+            if "newline_int" in idx_info_dict:
+                newline_int = idx_info_dict["newline_int"]
+                if self._newline_int != newline_int:
+                    logging.warning(
+                        f"Mismatch in newline_int, expected = {self._newline_int} but loaded {newline_int}"
+                    )
+
+            # test for version mismatch (useful to force recreation of index files)
+            idx_version = idx_info_dict.get("version", "0.0")
+            if __idx_version__ != idx_version:
+                raise RuntimeError(
+                    f"Version mismatch: Please delete existing '.{__idx_suffix__}' files. "
+                    f"Expected version = {__idx_version__}, but file version = {idx_version}. File path = {idx_fn}"
+                )
+        else:
+            raise ValueError(
+                f"Memory Map for {fn} is not found, missing one or more of files: {idx_fn}.{{.npy,.info}}"
+            )
+
+        return (mdata, midx)
+
+
+class JSONLMemMapDataset(TextMemMapDataset):
+    """
+    Memory-mapped iteration over a JSONL file.
+    """
+
+    def __init__(
+        self,
+        dataset_paths: List[str],
+        newline_int: Optional[int] = 10,
+        header_lines: Optional[int] = 0,
+        workers: Optional[int] = None,
+        tokenizer: Optional[Type["TokenizerSpec"]] = None,
+        sort_dataset_paths: Optional[bool] = True,
+        index_mapping_dir: Optional[str] = None,
+    ):
+        """
+        Args:
+            dataset_paths: list of JSONL file paths.
+            newline_int: ASCII code to use to interpret newlines in file.
+            header_lines: number of header lines in JSON files.
+            workers: number of workers to use for creating index files.
+            tokenizer: tokenizer to use to convert text to tokens.
+            sort_dataset_paths: whether to sort datasets by paths.
+            index_mapping_dir: directory to save the index mapping to.
+                If None, will write to the same folder as the dataset.
+        """
+        super().__init__(
+            dataset_paths=dataset_paths,
+            newline_int=newline_int,
+            header_lines=header_lines,
+            workers=workers,
+            tokenizer=tokenizer,
+            sort_dataset_paths=sort_dataset_paths,
+            index_mapping_dir=index_mapping_dir,
+        )
+
+    def _build_data_from_text(self, text):
+        """Return a dictionary of data based on a single JSON line."""
+        try:
+            record = json.loads(text)
+        except Exception as e:
+            logging.error(f"Exception: {e}")
+            logging.error(f"datapoint: {text}")
+            raise e
+        return record
+
+
+class OnlineSampleMapping:
```

Comment:
I think it makes more sense if TextMemMapDataset, JSONLMemMapDataset and OnlineSampleMapping are moved to a different file (maybe utils.py or maybe another file). Also, it might make sense to add a _ prefix before them to indicate that they're not to be instantiated by users